rmw/include/rmw/types.h

```diff
@@ -515,11 +515,73 @@ typedef struct RMW_PUBLIC_TYPE rmw_gid_s
+   * it should be taken consistently at the same point in the
+   * publishing process each time.
+   */
   rmw_time_point_value_t source_timestamp;
+  /// Time when the message was received by the subscription.
+  /**
+   * The exact point at which the timestamp is taken is not specified, but
+   * it should be taken consistently at the same point in the
+   * process of receiving a message each time.
+   */
   rmw_time_point_value_t received_timestamp;
+  /// Sequence number of the received message set by the publisher.
+  /**
+   * This sequence number is set by the publisher and therefore uniquely identifies
+   * a message when combined with the publisher GID.
+   * For long running applications, the sequence number might wrap around at some point.
+   *
+   * If the rmw implementation doesn't support sequence numbers, its value will be
+   * RMW_MESSAGE_INFO_SEQUENCE_NUMBER_UNSUPPORTED.
+   *
+   * Requirements:
+   *
+   * If `psn1` and `psn2` are the publication sequence numbers obtained by
+   * calls to `rmw_take*()`, where `psn1` was obtained in a call that happened before `psn2` and both
+   * sequence numbers are from the same publisher (i.e. also same publisher gid), then:
+   *
+   * - psn2 > psn1 (except in the case of a wrap around)
+   * - `psn2 - psn1 - 1` is the number of messages the publisher sent in the middle of both
+   *   received messages.
+   *   Those might have already been taken by other `rmw_take*()` calls that happened in between or lost.
+   *   `psn2 - psn1 - 1 = 0` if and only if the messages were sent by the publisher consecutively.
+   */
+  uint64_t publication_sequence_number;
+  /// Sequence number of the received message set by the subscription.
+  /**
+   * This sequence number is set by the subscription regardless of which
+   * publisher sent the message.
+   * For long running applications, the sequence number might wrap around at some point.
+   *
+   * If the rmw implementation doesn't support sequence numbers, its value will be
+   * RMW_MESSAGE_INFO_SEQUENCE_NUMBER_UNSUPPORTED.
+   *
+   * Requirements:
+   *
+   * If `rsn1` and `rsn2` are the reception sequence numbers obtained by
+   * calls to `rmw_take*()`, where `rsn1` was obtained in a call that happened before `rsn2`, then:
+   *
+   * - rsn2 > rsn1 (except in the case of a wrap around)
+   * - `rsn2 = rsn1 + 1` if and only if both `rmw_take*()` calls happened consecutively.
+   */
+  uint64_t reception_sequence_number;
+  /// Global unique identifier of the publisher that sent the message.
+  /**
+   * In rmw_cyclonedds, the "gid" currently is not global.
+   * i.e. the id printed by the local process for its publisher and the gid printed by a remote
+   * process may be different.
+   * gid continues being unique within a process though, and a
+   * (publication_sequence_number, publisher_gid) pair should still uniquely identify a message.
+   */
   rmw_gid_t publisher_gid;
```

Comment:
I would not mention cyclone directly. Instead I'd document the hard requirements and what we'd like in the future as a todo and then leave it at that.

My suggestion:

```suggestion
  /**
   * The identifier uniquely identifies the publisher for the local context, but
   * it will not necessarily be the same identifier given in other contexts or processes
   * for the same publisher.
   * Therefore the identifier will uniquely identify the publisher within your application
   * but may disagree about the identifier for that publisher when compared to another
   * application.
   * Even with this limitation, when combined with the publisher sequence number it can
   * uniquely identify a message within your local context.
   *
   * \todo In the future we want this to uniquely identify the publisher globally across
   *   contexts, processes, and machines.
   */
  rmw_gid_t publisher_gid;
```

I'm not sure if the limitation is per context or per process, but if it is per process then you could remove all the stuff about contexts, or I can update it again.

Reply:
Also, I don't know if we should mention that it can also potentially roll over? Not sure that would ever happen in practice, but it would be good to say what happens even if the answer is "we don't specify what happens in that case" (i.e. undefined behavior).

Reply:
> I don't know if we should mention that it can also potentially roll over? Not sure that would ever happen in practice

Yes, I think it could roll over (I'm not sure what exactly DDS vendors do), and some rmw implementations could use pseudo-random numbers as gid that could potentially collide (IIRC DDS doesn't do that).

Reply:
```suggestion
  /**
   * The identifier uniquely identifies the publisher for the local context, but
   * it will not necessarily be the same identifier given in other contexts or processes
   * for the same publisher.
   * Therefore the identifier will uniquely identify the publisher within your application
   * but may disagree about the identifier for that publisher when compared to another
   * application.
   * Even with this limitation, when combined with the publisher sequence number it can
   * uniquely identify a message within your local context.
   * Publishers gid generated by rmw implementations might at some point collide, in which
   * case it's not posible to distinguish from which message was the publisher.
   * The details of how GIDs are generated are rmw implementation dependent.
   *
   * \todo In the future we want this to uniquely identify the publisher globally across
   *   contexts, processes, and machines.
   */
  rmw_gid_t publisher_gid;
```

Reply:
Fixed in https://github.com/ros2/rmw/pull/318/commits/dcd3ea3ee6e958b6f575e7686c82b8ccfdbf1f07